### ISSUE TITLE
False positive TS2307 with typeCheck — path alias resolves in oxc_resolver but not in tsgolint type checker

### ISSUE THREAD

**Issue by imsyf** [oxc-project/tsgolint#1113]
**Setup:**
- `oxlint@1.75.0` + `oxlint-tsgolint@7.0.2001`
- `tsconfig.json` with `"paths": { "~/*": ["./*"] }` (no `baseUrl`)
- `tsc --noEmit` (TypeScript 7.0.2 via `@typescript/native`) resolves `~/` correctly — zero errors
- Config has `typeAware: true` and `typeCheck: true`

**Problem:**
Running `oxlint --type-aware --type-check` reports:
```
path/to/route.ts:4:27: error typescript(TS2307): Cannot find module '~/app/+constant/cache-tag' or its corresponding type declarations.
path/to/route.ts:5:33: error typescript(TS2307): Cannot find module '~/app/+lib/services/fe-cms/client' or its corresponding type declarations.
```
Both modules exist on disk and `tsc --noEmit` resolves them without error.

**Debug log (`OXC_LOG=debug`) shows the root cause:**
`oxc_resolver` (Rust) correctly reads `tsconfig.json`, parses `paths`, and resolves modules:
```
resolve{path=".../project/app/+lib/services/fe-cms" specifier="~/env/typed-env"}:
    oxc_resolver: options=ResolveOptions { alias: [], alias_fields: [], ... }
    ret=".../project/env/typed-env.ts"

find_tsconfig{path=.../project/env/typed-env.ts}:
    oxc_resolver::tsconfig_resolver:
    TsConfig {
        root: true, path: ".../project/tsconfig.json",
        compiler_options: CompilerOptions {
            base_url: None,
            paths: Some({"~/*": [".../*"], ...}),
            compiled_paths: Some(CompiledTsconfigPaths {
                wildcard_patterns: [
                    CompiledTsconfigPathPattern {
                        prefix: "~/", suffix: "",
                        targets: [Wildcard { prefix: ".../project/", suffix: "" }]
                    }
                ]
            })
        }
    }
```
`oxc_resolver` successfully maps `~/` → project root and resolves `~/env/typed-env` → `env/typed-env.ts`.

However, `tsgolint` (Go/typescript-go) still emits `TS2307` for `~/` imports, meaning the `compilerOptions.paths` are correctly parsed by `oxc_resolver` but not passed through to the `typescript-go` program for type-checking diagnostics.

This is distinct from the fixed #952 (which was about `extends`-based tsconfig chains). In this case, there's no `extends` — just a single root `tsconfig.json` with `paths`.

**Expected:**
`oxlint --type-check` should resolve `~/` the same way `tsc --noEmit` does, since both use `typescript-go`.

**Versions:**
- oxlint: 1.75.0
- oxlint-tsgolint: 7.0.2001
- @typescript/native: 7.0.2

**Comment by camc314** [comment]
Can you share the logs from oxlint-tsgolint so that I can understand which tsconfig is being used for that file?

**Comment by imsyf** [comment]
@camc314 Here are the tsgolint logs from `OXC_LOG=debug` focused on the failing file:

```
Got tsconfig for file path/to/app/.well-known/apple-app-site-association/route.ts: <none>
Done assigning files to programs. Total programs: 0. Unmatched files: 1
  Unmatched file: path/to/app/.well-known/apple-app-site-association/route.ts
```

The file is inside a `.well-known/` directory (dot-prefixed). tsgolint assigns it `<none>` tsconfig and 0 programs, so the type checker has no type info for it — hence the false TS2307.

In contrast, `oxc_resolver` (Rust) finds `tsconfig.json` for the same file without issue. And `tsc --noEmit` (TS 7.0.2) handles it correctly, suggesting the file lies within the tsconfig `include` patterns.

This looks like tsgolint's file-to-program assignment logic is skipping dot-prefixed directories.

**Comment by camc314** [comment]
hmm this is wierd - can you share the directory layout?

is there a tsconfig in a parent directory? what's the content of the tsconfig?

are only files in dot prefixed directories have this issue?

**Comment by imsyf** [comment]
@camc314 

**1. Directory layout:**
```
project/
├── tsconfig.json                    (include: ["**/*.ts"], no .* exclude)
├── cypress/tsconfig.json            (separate, for Cypress)
├── app/
│   ├── robots.ts                    ✅ assigned to tsconfig.json
│   ├── (redirect)/get-the-app/route.ts ✅ assigned to tsconfig.json
│   └── .well-known/…/route.ts       ❌ <none>
```

**2. tsconfig content:**
Root tsconfig has `"include": ["**/*.ts", …]`, no exclude of dot-dirs. `.well-known` should match fine but tsgolint returns `<none>` even with `--tsconfig tsconfig.json` passed explicitly.

**3. Only dot-prefixed? — Yes, confirmed.** I created a temp file in another dot-dir (`.husky/test.ts`) and got the same:
```
Got tsconfig for file …/.husky/test.ts: <none>
```
Non-dot files like `app/robots.ts`, `app/(redirect)/get-the-app/route.ts` all resolve to `tsconfig.json`.

**4. Minimal repro repository:** https://github.com/imsyf/tsgolint-dotdir-repro

To reproduce:
```sh
npm install
npx oxlint src/works.ts  # ✅ clean
npx oxlint .test/fails.ts # ❌ TS2307
```
The file in `.test/` (dot-prefixed) triggers false TS2307; same import from `src/` works fine.

**Comment by camc314** [comment]
Hi, thanks for the additional context to help me narrow down this issue.

I took a look, and the problem is not that `compilerOptions.paths` is being dropped by tsgolint.

TypeScript does not implicitly match dot-prefixed path components with:

```json
"include": ["**/*.ts"]
```

I confirmed this with the same TypeScript Go version: `tsgo --listFiles` includes `src/works.ts`, but not `.test/fails.ts`. Therefore, `tsc --noEmit` reports no errors because it never checks `.test/fails.ts`.

Since `.test/fails.ts` is not part of any configured TypeScript project, tsgolint assigns it to an inferred project. That inferred project does not inherit the root config’s `paths` mapping, which explains the TS2307 diagnostic.

The fix is to include the dot-directory explicitly:

```json
{
  "include": ["**/*.ts", "app/.well-known/**/*.ts"]
}
```

`--tsconfig` does not affect this because Oxlint’s type-aware linting currently ignores that option and discovers the appropriate TypeScript project for each file automatically.

I believe oxc_resolver resolving the tsconfig is a seperate issue: it discovers the nearest ancestor tsconfig and applies its `paths` mapping even when the importing file is not included in that TypeScript project. I have reproduced that independently and will investigate it on the resolver side.

Thanks!

**Comment by imsyf** [comment]
sorry @camc314, point of clarification

so all this time, TypeScript has silently ignored dot prefixed files?
could it be related to this? https://github.com/microsoft/TypeScript/issues/49555

**Comment by camc314** [comment]
> so all this time, TypeScript has silently ignored dot prefixed files?

yes - they simply weren't being rechecked

> could it be related to this? microsoft/TypeScript#49555

yes that looks related.


We've made type aware zero-config, this has the impact that issues such as this will be flagged if tsconfigs are not configured directly. (too add i think tho a behaviour (from Ts) is super confusing)
